app/krs.js

```diff
@@ -78,38 +85,50 @@ exports.provisionKey = function(req) {
         process.config.requesterAuth.clients[req.body.requesterId] !== req.body.requesterSecret) {
       throw utils.ErrorResponse(401, 'invalid requesterSecret');
     }
-    key.requesterId = req.body.requesterId;
   }
 
-  return key.saveQ()
-  .then(function() {
-    if (req.body.disableKRSEmail) {
-      // for API users who don't want their inboxes filling up with unread KRS backup key creation emails
-      return;
+  let masterKey = yield MasterKey.findOne({ customerId, coin });
+
+  if (!masterKey) {
+    masterKey = yield provisionMasterKey(coin, customerId);
+  }
+
+  key.masterKey = masterKey.xpub;
+  key.path = `m/${masterKey.keyCount}`;
+  key.xpub = utils.deriveChildKey(key.masterKey, key.path);
+
+  key.userEmail = req.body.userEmail;
+  key.notificationURL = req.body.notificationURL;
+
+  key.custom = req.body.custom || {};
+  key.custom.created = new Date();
+
+  yield key.save();
+
+  yield masterKey.update({ $inc: { keyCount: 1 } });
+
+  if (!req.body.disableKRSEmail) {
+    try {
+      yield utils.sendMailQ(
+        key.userEmail,
+        'Information about your BitGo backup key',
+        'newkeytemplate',
+        {
+          xpub: key.xpub,
+          servicename: process.config.name,
+          serviceurl: process.config.serviceurl,
+          adminemail: process.config.adminemail,
+          useremail: key.userEmail
+        });
+    } catch (e) {
+      throw utils.ErrorResponse(503, 'Problem sending email');
     }
-    return utils.sendMailQ(
-      userEmail,
-      "Information about your backup key",
-      "newkeytemplate",
-      {
-        xpub: xpub,
-        servicename: process.config.name,
-        serviceurl: process.config.serviceurl,
-        adminemail: process.config.adminemail,
-        useremail: userEmail
-      }
-    )
-    .catch(function(e) {
-      throw utils.ErrorResponse(503, "Problem sending email");
-    });
-  })
-  .then(function() {
-    return notifyEndpoint(key, 'created');
-  })
-  .then(function() {
-    return key;
-  });
-};
+  }
+
+  yield notifyEndpoint(key, 'created');
```

Comment:
Maybe do the check for if(notifyEndpoint == null) here, for readability and so we don't go to the trouble of generating HMAC and then just breaking without using it. Or you could turn notifyEndpoint into just notify and have both the email logic and the notifyEndpoint logic in there, switched on the flags